### PR TITLE
Implement Blockwise Op to vectorize existing Ops

### PR DESCRIPTION
**Pull request by ricardoV94**
Critical for https://github.com/pymc-devs/pymc/issues/5383

This PR was done almost completely from scratch, although the general principles and some of the code were obviously inspired by https://github.com/aesara-devs/aesara/pull/1215

For fairness, all authors of the commits in that PR were added as co-authors here.

CC @purna135

Some design questions:
1. Do we want to make Blockwise the "default" form of an Op (in terms of rewrites), and only later specialize if they were not needed?  **Decided yes**
2. Do we want to fuse consecutive Blockwises with the same batched dims, similarly to how we fuse consecutive Elemwises? **Not for now**
3. Can Rop be implemented just as the batched core Rop, like L_op?

Important follow-up:
1. Dispatch to JAX and Numba
2. ~~Create a PyTensor "vectorize" function that vectorizes the whole graph, similar to what the `L_op` method has to do~~ Included as the last commit

```python
from pytensor.graph import vectorize

# Original graph
x = pt.vector("x")
y = pt.exp(x) / pt.sum(pt.exp(x))

new_x = pt.matrix("new_x")
[new_y] = vectorize([y], {x: new_x})

fn = pytensor.function([new_x], new_y)
fn([[0, 1, 2], [2, 1, 0]])
# array([[0.09003057, 0.24472847, 0.66524096],
#        [0.66524096, 0.24472847, 0.09003057]])
```

Important to test:
1. Whether this works with Ops with inner graphs (Scan, Composite)

### CODE REVIEW

File: pytensor/tensor/blockwise.py [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+from pytensor.tensor.elemwise import DimShuffle, Elemwise
+from pytensor.tensor.shape import shape_padleft
+from pytensor.tensor.type import continuous_dtypes, discrete_dtypes, tensor
+from pytensor.tensor.var import TensorVariable, as_tensor_variable
+
+
+# TODO: Implement vectorize helper to batch whole graphs (similar to what Blockwise does for the grad)
+
+# TODO: Add github link
+# Copied verbatim from numpy.lib.function_base
```
Comment (ricardoV94) [inline pytensor/tensor/blockwise.py:21]:
```suggestion
# Copied verbatim from numpy.lib.function_base
# https://github.com/numpy/numpy/blob/f2db090eb95b87d48a3318c9a3f9d38b67b0543c/numpy/lib/function_base.py#L1999-L2029
```

File: pytensor/tensor/blockwise.py [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+            if missing_batch_ndims:
+                inp = shape_padleft(inp, missing_batch_ndims)
+            batched_inputs.append(inp)
+
+            if not sig:
+                batch_shapes.append(inp.type.shape)
+            else:
+                batch_shapes.append(inp.type.shape[: -len(sig)])
+
+        def get_most_specialized_batch_shape(
```
Comment (ferrine) [inline pytensor/tensor/blockwise.py:189]: I remember I saw similar code in elemwise, can we avoid duplication there?

File: pytensor/tensor/elemwise.py [hunk 3]
```diff
@@ -1759,3 +1746,37 @@ def _get_vector_length_Elemwise(op, var):
...
+    ]
+    return DimShuffle(input_broadcastable, new_order).make_node(x)
+
+
+@_vectorize_node.register(CAReduce)
+def vectorize_careduce(op: CAReduce, node: Apply, x: TensorVariable) -> Apply:
+    batched_ndims = x.type.ndim - node.inputs[0].type.ndim
+    if not batched_ndims:
+        return node.op.make_node(x)
+    axes = op.axis
```
Comment (ferrine) [inline pytensor/tensor/elemwise.py:1773]: are those axis always in norm form?

File: pytensor/tensor/nlinalg.py [hunk 4]
```diff
@@ -112,6 +95,8 @@ class MatrixInverse(Op):
     """
 
     __props__ = ()
+    gufunc_signature = "(m,m)->(m,m)"
+    gufunc_spec = ("numpy.linalg.inv", 1, 1)
```
Comment (ferrine) [inline pytensor/tensor/nlinalg.py:99]: what are `1, 1`? inputs and outputs?

File: pytensor/tensor/nlinalg.py [hunk 5]
```diff
@@ -228,7 +215,7 @@ def __str__(self):
         return "Det"
 
 
-det = Det()
+det = Blockwise(Det())
```
Comment (ferrine) [inline pytensor/tensor/nlinalg.py:218]: is there a better way to always have blockwise ops?

File: pytensor/tensor/rewriting/linalg.py [hunk 6]
```diff
@@ -17,16 +19,40 @@
...
+        [inp] = node.inputs
+        ndims = inp.type.ndim
+        if ndims < 2:
+            return False
+        transpose_order = tuple(range(ndims - 2)) + (ndims - 1, ndims - 2)
+        return cast(bool, node.op.new_order == transpose_order)
+    return False
+
+
+def T(x: TensorVariable) -> TensorVariable:
```
Comment (ferrine) [inline pytensor/tensor/rewriting/linalg.py:39]: numpy transpose does a different thing

image

File: pytensor/tensor/rewriting/linalg.py [hunk 7]
```diff
@@ -37,87 +63,98 @@ def inv_as_solve(fgraph, node):
...
+                return [T(solve(x, T(l)))]
             else:
-                return [solve(x.T, l.T).T]
+                return [T(solve(T(x), T(l)))]
 
 
 @register_stabilize
 @register_canonicalize
-@node_rewriter([Solve])
+@node_rewriter([Blockwise])
```
Comment (ferrine) [inline pytensor/tensor/rewriting/linalg.py:86]: I think there should be a better design to register blockwise rewrites of this kind

File: pytensor/tensor/rewriting/linalg.py [hunk 8]
```diff
@@ -37,87 +63,98 @@ def inv_as_solve(fgraph, node):
...
-                if A.owner.op.lower:
-                    return [Solve(assume_a="sym", lower=True)(A, b)]
-                else:
-                    return [Solve(assume_a="sym", lower=False)(A, b)]
             if (
                 A.owner
-                and isinstance(A.owner.op, DimShuffle)
-                and A.owner.op.new_order == (1, 0)
+                and isinstance(A.owner.op, Blockwise)
+                and isinstance(A.owner.op.core_op, Cholesky)
```
Comment (ferrine) [inline pytensor/tensor/rewriting/linalg.py:99]: This becomes messy if there are blockwise ops and non blockwise ops under consideration.

**Review by ferrine — COMMENTED** [review]
I see issues with the design approach. The rewrites become more and more involved and should take blockwise and non blockwise ops in account. I think a discussion over this issue is required before any steps forward.

File: pytensor/tensor/elemwise.py (hunk 3 above)
Comment (ricardoV94) [reply pytensor/tensor/elemwise.py:1773]: Should check

File: pytensor/tensor/nlinalg.py (hunk 4 above)
Comment (ricardoV94) [reply pytensor/tensor/nlinalg.py:99]: It's the number of inputs and outputs, just following the same convetion of `nfunc_spec` for Scalar Ops

File: pytensor/tensor/nlinalg.py (hunk 5 above)
Comment (ricardoV94) [reply pytensor/tensor/nlinalg.py:218]: This is the same as `Elemwise` always creating from `Scalar` when you import PyTensor

File: pytensor/tensor/rewriting/linalg.py (hunk 6 above)
Comment (ricardoV94) [reply pytensor/tensor/rewriting/linalg.py:39]: This is just a helper function for the rewrites in this section that only transpose the last two dimensions. Makes the rewrites easier to read

File: pytensor/tensor/rewriting/linalg.py (hunk 7 above)
Comment (ricardoV94) [reply pytensor/tensor/rewriting/linalg.py:86]: This is the same issue for Elemwises that must be parametrized (created at runtime)

File: pytensor/tensor/rewriting/linalg.py (hunk 8 above)
Comment (ricardoV94) [reply pytensor/tensor/rewriting/linalg.py:99]: Rewrites now never work for non-blockwise. We expect users to always work with blockwise (the same way we always expect users to work with Elemwise and not scalar Ops)

**Comment by ricardoV94** [comment]
@ferrine I think the sanest thing is for rewrites to only target Blockwise graphs, the same way that our rewrites only target Elemwise graphs and never the equivalent Scalar graphs (see some discussion in #107).

Users would now always work with Blockwise (that's what we expose by default when you do pt.foo), the same way they always work with Elemwise. During compilation we remove Blockwise when unnecessary the same way that #107 suggests to remove Elemwise when unnecessary.

What problem do you see with this?

File: pytensor/tensor/rewriting/linalg.py (hunk 6 above)
Comment (ferrine) [reply pytensor/tensor/rewriting/linalg.py:39]: I think it should not be there, easy to confuse with numpy transpose and it does it differntly

**Comment by ferrine** [comment]
Let's get more reviews

File: pytensor/tensor/rewriting/linalg.py (hunk 6 above)
Comment (ricardoV94) [reply pytensor/tensor/rewriting/linalg.py:39]: The point was to make the rewrites changes here to track/ review. What before was var.T now is T(var) because we keep working only on the last 2 dimensions. All rewrites work for batches of matrices the linear algebra equivalence still comes from the matrix case.

Happy to add a leading underscore, or call it something more verbose like matrix_T

File: pytensor/tensor/slinalg.py [hunk 9]
```diff
@@ -204,15 +137,24 @@ class SolveBase(Op):
...
+        b_ndim,
     ):
         self.lower = lower
         self.check_finite = check_finite
+        assert b_ndim in (1, 2)
+        self.b_ndim = b_ndim
+        if b_ndim == 1:
+            self.gufunc_signature = "(m,m),(m)->(m)"
+        else:
+            self.gufunc_signature = "(m,m),(m,m)->(m,m)"
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:157]:
```suggestion
            self.gufunc_signature = "(m,m),(m,n)->(m,n)"
```

File: pytensor/tensor/random/op.py [hunk 10]
```diff
@@ -428,3 +429,8 @@ class DefaultGeneratorMakerOp(AbstractRNGConstructor):
 
 
 default_rng = DefaultGeneratorMakerOp()
+
+
+# RandomVariables are vectorized on the parameters by default.
+# RNG, size and dtype can't be vectorized, but the Op will raise if the wrong input type is passed
+_vectorize_node.register(RandomVariable, vectorize_not_needed)
```
Comment (ricardoV94) [inline pytensor/tensor/random/op.py:436]: TODO: Consider size

File: pytensor/tensor/nlinalg.py [hunk 11]
```diff
@@ -200,6 +185,8 @@ class Det(Op):
     """
 
     __props__ = ()
+    gufunc_signature = "(m,m)->()"
```
Comment (jessegrabowski) [inline pytensor/tensor/nlinalg.py:188]: Opening a comment to discuss the signatures, since that was a major bone of contention in the last meeting.

I understood @aseyboldt 's position essentially to be (please correct me if I'm wrong) that the signatures represent only partial information, so they more confusing than helpful in the overall design of an `Op`. For example, you could not use this signature to construct a `numba` function signature from an `Op` alone. Secondly, the signatures don't actually contain the information needed by `Blockwise`. Instead, `Blockwise` has to reconstruct what it needs from the signatures, which are only useful in this one context anyway (see point 1), so their purpose ends up obscured. 

My two cents: I like that it matches the numpy syntax. Perhaps the name could be changed to make it more obvious that these are information specifically to be used by blockwise? Like `blockwise_signature`, or `blockwise_core_sig`? 

There was also a discussion about whether this information should live in the Ops, or whether it should be with Blockwise. I think it conceptually makes sense the way it is -- it's a more OOP design pattern this way, since Blockwise "asks" the Op what how it should interact with it.

Comment (ricardoV94) [reply pytensor/tensor/nlinalg.py:188]: Noting that @ferrine explored more rich signatures in https://github.com/pymc-devs/pytensor/pull/143

I do think that although they can help they will never be "feature complete" because they would probably require being turing complete :)

File: pytensor/tensor/elemwise.py (hunk 3 above)
Comment (ricardoV94) [reply pytensor/tensor/elemwise.py:1773]: They can be negative, although calling `pt.sum()` normalizes them. I made the dispatch work with both cases.

Comment (ricardoV94) [reply pytensor/tensor/elemwise.py:1773]: Actually `CAReduce.make_node` recreates the Op with positive axis, so there can't be negative axis ever

File: pytensor/tensor/blockwise.py (hunk 2 above)
Comment (ricardoV94) [reply pytensor/tensor/blockwise.py:189]: refactored

**Comment by ricardoV94** [comment]
Last commits introduces a vectorize utility. This should be the PyTensor equivalent to `np.vectorize` or `jax.vectorize`. 

If people disagree too much with the API I will drop this commit so we can discuss it elsewhere. I would like to get the Blockwise functionality merged regardless of what the final `vectorize` may look like, because this functionality is sorely missing.

**Comment by twiecki** [comment]
I think in pymc4 we tried to use functionality like this to sample multiple chains in parallel (in TF). Sounds like we could just take a pymc model logp and vectorize it along the dim of chains. Now the question is whether that would provide any benefits.

**Comment by ricardoV94** [comment]
> I think in pymc4 we tried to use functionality like this to sample multiple chains in parallel (in TF). Sounds like we could just take a pymc model logp and vectorize it along the dim of chains. Now the question is whether that would provide any benefits.

I am not sure what numpyro vectorize does, but that sounds a bit strange? The sampler probably wants to go different tree_depths for different chains / points?

I see this as being more useful for actual model building, where you define the core case and then vectorize batched dims. 

Also useful for marginalization of discrete variables via enumeration that `pymc_experimental.marginal_model` does.

Might also be useful for stuff like `compute_log_likelihood` and prior and posterior predictive sampling where we now resort to iterating over chains/draws, although I am not sure what cases would actually be faster.

**Comment by twiecki** [comment]
> I am not sure what numpyro vectorize does, but that sounds a bit strange? The sampler probably wants to go different tree_depths for different chains / points?

Right, for NUTS it's a pain, as @ColCarroll can attest to. I think that's where some of the new sampler developments might come in handy by Matt Hoffman et al.
 
> I see this as being more useful for actual model building, where you define the core case and then vectorize batched dims.
> 
> Also useful for marginalization of discrete variables via enumeration that `pymc_experimental.marginal_model` does.
> 
> Might also be useful for stuff like `compute_log_likelihood` and prior and posterior predictive sampling where we now resort to iterating over chains/draws, although I am not sure what cases would actually be faster.

Interesting potential applications.

**Comment by ricardoV94** [comment]
> Merge?

I think so. Nothing like trying it out there to see if the design is good
